src/Runtime/Distributions/Automata/Automaton.Builder.cs

```diff
@@ -0,0 +1,810 @@
+                {
+                    iterator.Remove();
+                }
+
+                this.states.RemoveAt(stateIndex);
+
+                for (var i = 0; i < this.states.Count; ++i)
+                {
+                    for (var iterator = this[i].TransitionIterator; iterator.Ok; iterator.Next())
+                    {
+                        var transition = iterator.Value;
+                        if (transition.DestinationStateIndex > stateIndex)
+                        {
+                            transition.DestinationStateIndex -= 1;
+                            iterator.Value = transition;
+                        }
+                        else if (transition.DestinationStateIndex == stateIndex)
+                        {
+                            iterator.Remove();
+                        }
+                    }
+                }
+            }
+
+            /// <summary>
+            /// Removes a set of states from the automaton where the set is defined by labels not matching
+            /// the <paramref name="removeLabel"/>.
+            /// </summary>
+            /// <param name="labels">State labels</param>
+            /// <param name="removeLabel">Label which marks states which should be deleted</param>
+            public int RemoveStates(bool[] labels, bool removeLabel)
+            {
+                var oldToNewStateIdMapping = new int[this.states.Count];
+                var newStateId = 0;
+                var deadStateCount = 0;
+                for (var stateId = 0; stateId < this.states.Count; ++stateId)
+                {
+                    if (labels[stateId] != removeLabel)
+                    {
+                        oldToNewStateIdMapping[stateId] = newStateId++;
+                    }
+                    else
+                    {
+                        oldToNewStateIdMapping[stateId] = -1;
+                        ++deadStateCount;
+                    }
+                }
+
+                this.StartStateIndex = oldToNewStateIdMapping[this.StartStateIndex];
+                if (this.StartStateIndex == -1)
+                {
+                    // Cannot reach any end state from the start state => the automaton is zero everywhere
+                    this.SetToZero();
+                    return deadStateCount;
+                }
+
+                for (var i = 0; i < this.states.Count; ++i)
+                {
+                    var newId = oldToNewStateIdMapping[i];
+                    if (newId == -1)
+                    {
+                        // remove all transitions
+                        for (var iterator = this[i].TransitionIterator; iterator.Ok; iterator.Next())
+                        {
+                            iterator.Remove();
+                        }
+
+                        continue;
+                    }
+
+                    Debug.Assert(newId <= i);
```

Comment:
Debug.Assert

Reply:
It is in there. What is wrong?

Reply:
My question is why you would want to include the Debug.Assert in the repo.  If it's there for a specific reason, I think it should be commented.

Reply:
Well, Asserts are there to check for some invariants which should be checked (or at least stated). For performance reasons I'd like that to be done only in debug. They are essentially free in production builds, but help with refactoring.